macros/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+                        FnArg::Captured(ref arg) => match arg.ty {
+                            Type::Reference(ref r) => {
+                                r.lifetime.is_none() && r.mutability.is_none()
+                            }
+                            _ => false,
+                        },
+                        _ => false,
+                    }
+                    && f.decl.generics.params.is_empty()
+                    && f.decl.generics.where_clause.is_none()
+                    && f.decl.variadic.is_none()
+                    && match f.decl.output {
+                        ReturnType::Default => false,
+                        ReturnType::Type(_, ref ty) => match **ty {
+                            Type::Never(_) => true,
+                            _ => false,
+                        },
+                    },
+                "`#[exception(HardFault)]` function must have signature `fn(&ExceptionFrame) -> !`"
+            );
+
+            assert!(
+                args.second.is_none(),
+                "`#[exception(HardFault)]` takes no additional arguments"
+            );
+
+            let arg = match f.decl.inputs[0] {
+                FnArg::Captured(ref arg) => arg,
+                _ => unreachable!(),
+            };
+
+            let pat = &arg.pat;
+
+            quote!(
+                #[export_name = "UserHardFault"]
+                pub unsafe extern "C" fn #ident(#arg) -> ! {
+                    extern crate cortex_m_rt;
+
+                    // further type check of the input argument
+                    let #pat: &cortex_m_rt::ExceptionFrame = #pat;
+
+                    #(#stmts)*
+                }
+            ).into()
+        }
+        Exception::Other => {
+            assert!(
+                f.constness.is_none()
+                    && f.vis == Visibility::Inherited
+                    && f.unsafety.is_none()
+                    && f.abi.is_none()
+                    && f.decl.inputs.is_empty()
+                    && f.decl.generics.params.is_empty()
+                    && f.decl.generics.where_clause.is_none()
+                    && f.decl.variadic.is_none()
+                    && match f.decl.output {
+                        ReturnType::Default => true,
+                        ReturnType::Type(_, ref ty) => match **ty {
+                            Type::Tuple(ref tuple) => tuple.elems.is_empty(),
+                            _ => false,
+                        },
+                    },
+                "`#[exception]` functions must have signature `fn()`"
+            );
+
+            if let Some(second) = args.second {
+                let ty = second.ty;
+                let expr = second.expr;
+                let state = second.ident;
+
+                quote!(
+                    #[export_name = #name_s]
+                    pub fn #ident() {
+                        extern crate cortex_m_rt;
+
+                        cortex_m_rt::Exception::#name;
```

Comment:
@adamgreig this is reason why an hypothetical `#[interrupt]` attribute would need the device crate name in its argument. We want to check that the interrupt exists and emit a compiler error if it doesn't and for that we need the full path to the `Interrupt` variant: e.g. `stm32f103xx::Interrupt::EXTI0`.

Reply:
Makes sense. I guess we could require that whatever is passed in (to some future `#[interrupt]`) is resolvable to an `Interrupt`, so it could be either a full path, or the user could `use stm32f103::Interrupt::*` and then just call `#[interrupt(EXTI0)]`?

Reply:
>  the user could use stm32f103::Interrupt::* and then just call #[interrupt(EXTI0)]?

If we omit the full path that would kind of defeat the purpose of the check. For example, this would compile:

``` rust
struct InterruptThatDoesntExist;

#[interrupt(InterruptThatDoesntExist)]
fn handler() { /* .. */ }
```

Though thinking about it more having *the user* enter the name of the device crate makes it possible to nullify the check. For example,

``` rust
extern crate stm32f103xx; // real device crate

mod stm32f1337 {
    enum Interrupt {
        InterruptThatDoesntExist;
    }
}

#[interrupt(stm32f1337, InterruptThatDoesntExist)]
fn handler() { /* .. */ }
```